File: src-tauri/src/commands/channel.rs
use crate::models::result::CommandResult;
use crate::services::channel_service::{
    self, Channel, CreateChannelPayload, DeleteChannelData, UpdateChannelPayload,
};

#[tauri::command]
pub async fn list_channels() -> CommandResult<Vec<Channel>> {
    match channel_service::list_channels().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn add_channel(payload: CreateChannelPayload) -> CommandResult<Channel> {
    match channel_service::add_channel(payload).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn update_channel(payload: UpdateChannelPayload) -> CommandResult<Channel> {
    match channel_service::update_channel(payload).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn delete_channel(id: String) -> CommandResult<DeleteChannelData> {
    match channel_service::delete_channel(id).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

File: src/main/ipc/channel.ts
import { ipcMain } from "electron";
import { IPC } from "../../shared/ipc-channels.js";
import { ok, err } from "../models/result.js";
import { toAppError } from "./utils.js";
import {
  listChannels,
  addChannel,
  updateChannel,
  deleteChannel,
} from "../services/channel-service.js";
import type {
  CreateChannelPayload,
  UpdateChannelPayload,
} from "../services/channel-service.js";

export function registerChannelHandlers(): void {
  ipcMain.handle(IPC.LIST_CHANNELS, async () => {
    try { return ok(await listChannels()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.ADD_CHANNEL, async (_event, payload: CreateChannelPayload) => {
    try { return ok(await addChannel(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.UPDATE_CHANNEL, async (_event, payload: UpdateChannelPayload) => {
    try { return ok(await updateChannel(payload)); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.DELETE_CHANNEL, async (_event, payload: { id: string }) => {
    try { return ok(await deleteChannel(payload.id)); }
    catch (error) { return err(toAppError(error)); }
  });
}
